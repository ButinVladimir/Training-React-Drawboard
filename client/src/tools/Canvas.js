export const DEFAULT_DELTA_X = 0.0;
export const DEFAULT_DELTA_Y = 0.0;
export const DEFAULT_ZOOM = 1.0;
export const DEFAULT_ROTATION = 0.0;
export const DEFAULT_WIDTH = 500;
export const DEFAULT_HEIGHT = 500;
export const DEFAULT_CANVAS_COLOR = '#fff';
export const BORDER_COLOR = '#999';
export const BORDER_WIDTH = 3;

class Canvas {
  constructor() {
    this.deltaX = DEFAULT_DELTA_X;
    this.deltaY = DEFAULT_DELTA_Y;
    this.zoom = DEFAULT_ZOOM;
    this.rotation = DEFAULT_ROTATION;

    this.containerPositionX = 0;
    this.containerPositionY = 0;

    this.screenCanvasValue = null;
    this.screenContext = null;

    this.htmlCanvasElement = document.createElement('canvas');
    this.htmlCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.htmlCanvasElement.setAttribute('height', DEFAULT_HEIGHT);

    this.htmlCanvasContext = this.htmlCanvasElement.getContext('2d');
    this.htmlCanvasContext.fillStyle = DEFAULT_CANVAS_COLOR;
    this.htmlCanvasContext.fillRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.bufferCanvasElement = document.createElement('canvas');
    this.bufferCanvasElement.setAttribute('width', DEFAULT_WIDTH);
    this.bufferCanvasElement.setAttribute('height', DEFAULT_HEIGHT);
    this.bufferCanvasContext = this.bufferCanvasElement.getContext('2d');
  }

  get screenCanvas() {
    return this.screenCanvasValue;
  }

  set screenCanvas(value) {
    this.screenCanvasValue = value;
    this.screenContext = value.getContext('2d');
  }

  outputCompletely() {
    this.screenContext.resetTransform();
    this.screenContext.clearRect(0, 0, this.screenCanvas.width, this.screenCanvas.height);

    this.show();
  }

  changeDelta(deltaX, deltaY) {
    this.hide();
    this.deltaX = deltaX;
    this.deltaY = deltaY;
    this.show();
  }

  changeZoom(zoom) {
    this.hide();
    this.zoom = zoom;
    this.show();
  }

  changeRotation(rotation) {
    this.hide();
    this.rotation = rotation;
    this.show();
  }

  applyTransformations() {
    this.screenContext.resetTransform();
    this.screenContext.translate(this.deltaX, this.deltaY);
    this.screenContext.rotate(this.rotation);
    this.screenContext.scale(this.zoom, this.zoom);
  }

  hide() {
    this.applyTransformations();

    this.screenContext.clearRect(
      -2 * BORDER_WIDTH, -2 * BORDER_WIDTH,
      DEFAULT_WIDTH + 4 * BORDER_WIDTH,
      DEFAULT_HEIGHT + 4 * BORDER_WIDTH,
    );
  }

  show() {
    this.applyTransformations();

    this.screenContext.strokeStyle = BORDER_COLOR;
    this.screenContext.lineWidth = BORDER_WIDTH;
    this.screenContext.strokeRect(0, 0, DEFAULT_WIDTH, DEFAULT_HEIGHT);

    this.screenContext.drawImage(this.htmlCanvasElement, 0, 0);
  }

  clearBuffer() {
    this.bufferCanvasContext.resetTransform();
    this.bufferCanvasContext.clearRect(
      0,
      0,
      this.bufferCanvasElement.width,
      this.bufferCanvasElement.height,
    );
  }

  updateAndShowBuffer(updateCb, ...args) {
    this.clearBuffer();

    this.bufferCanvasContext.scale(1 / this.zoom, 1 / this.zoom);
    this.bufferCanvasContext.rotate(-this.rotation);
    this.bufferCanvasContext.translate(-this.deltaX, -this.deltaY);
    this.bufferCanvasContext.translate(-this.containerPositionX, -this.containerPositionY);

    updateCb(this.bufferCanvasContext, ...args);

    this.show();
    this.screenContext.drawImage(this.bufferCanvasElement, 0, 0);
  }

  appendBuffer() {
    this.htmlCanvasContext.resetTransform();
    this.htmlCanvasContext.drawImage(this.bufferCanvasElement, 0, 0);
    this.clearBuffer();

    this.show();
  }
}

export default Canvas;
