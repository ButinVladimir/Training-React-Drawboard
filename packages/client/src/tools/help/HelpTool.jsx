import React from 'react';
import BaseTool from '../BaseTool';
import HelpToolOptions from '../../components/tools/help/HelpToolOptionsContainer';

const DEFAULT_TOOL_KEY = 'toolKey';

class HelpTool extends BaseTool {
  constructor(canvas, toolState, stateHandler) {
    super(canvas, toolState, stateHandler);

    this.toolKey = DEFAULT_TOOL_KEY;

    this.onRestore = this.onRestore.bind(this);
  }

  onRestore() {
    this.stateHandler.restore(this.toolState);
  }

  get optionsElement() {
    return (
      <HelpToolOptions
        key={this.toolKey}
        toolState={this.toolState}
        onRestore={this.onRestore}
      />
    );
  }
}

export default HelpTool;
