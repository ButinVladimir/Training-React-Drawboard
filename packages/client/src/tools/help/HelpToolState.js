import BaseToolState from '../BaseToolState';

class HelpToolState extends BaseToolState {
  constructor() {
    super();

    this.anchorX = 0;
    this.anchorY = 0;

    this.clientX = 0;
    this.clientY = 0;

    this.speed = 0;
  }

  restore() { // eslint-disable-line class-methods-use-this,no-unused-vars
    throw new Error('Method is not implemented');
  }

  setDefaultSpeed() { // eslint-disable-line class-methods-use-this,no-unused-vars
    throw new Error('Method is not implemented');
  }

  onMouseDown(event) {
    this.anchorX = event.clientX;
    this.anchorY = event.clientY;

    this.clientX = event.clientX;
    this.clientY = event.clientY;
  }

  onMouseMove(event) {
    this.clientX = event.clientX;
    this.clientY = event.clientY;
  }

  onMouseUp() {
    this.clientX = this.anchorX;
    this.clientY = this.anchorY;
  }

  serialize() { // eslint-disable-line class-methods-use-this
    return null;
  }

  static deserialize() {
    return null;
  }
}

export default HelpToolState;
