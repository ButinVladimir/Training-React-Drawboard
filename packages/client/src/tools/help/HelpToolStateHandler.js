import BaseToolStateHandler from '../BaseToolStateHandler';

class HelpToolStateHandler extends BaseToolStateHandler {
  onMouseMove(toolState) {
    this.applyState(null, null, toolState);
  }

  onMouseUp(toolState) { // eslint-disable-line class-methods-use-this
    toolState.onMouseUp();

    return null;
  }

  applyState(canvasContext, viewState, toolState) {
    this.canvas.hide();
    this.changeViewState(toolState);
    this.canvas.show();
  }

  changeViewState(toolState) { // eslint-disable-line class-methods-use-this,no-unused-vars
    throw new Error('Method is not implemented');
  }

  restore(toolState) {
    toolState.restore();
    this.applyState(null, null, toolState);
  }
}

export default HelpToolStateHandler;
