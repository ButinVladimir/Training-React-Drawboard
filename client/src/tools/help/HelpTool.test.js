import { shallow } from 'enzyme';
import HelpTool from './HelpTool';
import HelpToolState from './HelpToolState';
import HelpToolStateHandler from './HelpToolStateHandler';
import Canvas from '../Canvas';

describe('HelpTool', () => {
  it('is instantiated properly', () => {
    const canvas = new Canvas();
    const toolState = new HelpToolState();
    const stateHandler = new HelpToolStateHandler(canvas);
    const helpTool = new HelpTool(canvas, toolState, stateHandler);

    expect(helpTool).toBeInstanceOf(HelpTool);
    expect(helpTool.canvas).toBe(canvas);
    expect(helpTool.toolState).toBe(toolState);
    expect(helpTool.stateHandler).toBe(stateHandler);
    expect(helpTool.toolKey).toBe('toolKey');
  });

  it('restores its state', () => {
    const canvas = new Canvas();
    const toolState = new HelpToolState();
    const stateHandler = {
      restore: jest.fn(),
    };
    const helpTool = new HelpTool(canvas, toolState, stateHandler);

    helpTool.onRestore();

    expect(stateHandler.restore).toHaveBeenCalledTimes(1);
    expect(stateHandler.restore).toHaveBeenCalledWith(toolState);
  });

  it('returns options element', () => {
    const canvas = new Canvas();
    const toolState = new HelpToolState();
    const stateHandler = new HelpToolStateHandler(canvas);
    const helpTool = new HelpTool(canvas, toolState, stateHandler);

    const options = shallow(helpTool.optionsElement);

    expect(options).toMatchSnapshot();
  });
});
