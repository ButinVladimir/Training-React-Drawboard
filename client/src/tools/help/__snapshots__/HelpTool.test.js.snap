// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpTool returns options element 1`] = `
<HelpToolOptions
  onChangeSpeed={[Function]}
  onRestore={[Function]}
  onSetDefaultSpeed={[Function]}
  parsedSpeed={0}
  speed={0}
/>
`;
